src/ripple/overlay/impl/PeerImp.cpp

```diff
@@ -1127,221 +1114,260 @@ PeerImp::onMessage(std::shared_ptr<protocol::TMCluster> const& m)
         fee_ = Resource::feeBadData;
         JLOG(p_journal_.warn()) << msg;
     };
 
-    if (m->hops() > csHopLimit)
-        return badData("Invalid hops: " + std::to_string(m->hops()));
-    if (m->peerchain_size() > csHopLimit)
-        return badData("Invalid peer chain");
+    // Verify relays
+    if (m->relays() > relayLimit)
+        return badData("Invalid relays");
 
-    // Reply with shard info we may have
-    if (auto shardStore = app_.getShardStore())
+    // Verify peer chain
+    // The peer chain should not contain this node's public key
+    // nor the public key of the sending peer
+    std::set<PublicKey> pubKeyChain;
+    pubKeyChain.insert(app_.nodeIdentity().first);
+    pubKeyChain.insert(publicKey_);
+
+    auto const peerChainSz{m->peerchain_size()};
+    if (peerChainSz > 0)
     {
-        fee_ = Resource::feeLightPeer;
-        auto shards{shardStore->getCompleteShards()};
-        if (!shards.empty())
-        {
-            protocol::TMPeerShardInfo reply;
-            reply.set_shardindexes(shards);
+        if (peerChainSz > relayLimit)
+            return badData("Invalid peer chain size");
 
-            if (m->has_lastlink())
-                reply.set_lastlink(true);
+        if (peerChainSz + m->relays() > relayLimit)
+            return badData("Invalid relays and peer chain size");
 
-            if (m->peerchain_size() > 0)
-            {
-                for (int i = 0; i < m->peerchain_size(); ++i)
-                {
-                    if (!publicKeyType(makeSlice(m->peerchain(i).nodepubkey())))
-                        return badData("Invalid peer chain public key");
-                }
-
-                *reply.mutable_peerchain() = m->peerchain();
-            }
+        for (int i = 0; i < peerChainSz; ++i)
+        {
+            auto const slice{makeSlice(m->peerchain(i).publickey())};
 
-            send(std::make_shared<Message>(reply, protocol::mtPEER_SHARD_INFO));
+            // Verify peer public key
+            if (!publicKeyType(slice))
+                return badData("Invalid peer public key");
 
-            JLOG(p_journal_.trace()) << "Sent shard indexes " << shards;
+            // Verify peer public key is unique in the peer chain
+            if (!pubKeyChain.emplace(slice).second)
+                return badData("Invalid peer public key");
         }
     }
 
-    // Relay request to peers
-    if (m->hops() > 0)
+    // Reply with shard info this node may have
+    if (auto shardStore = app_.getShardStore())
     {
-        fee_ = Resource::feeMediumBurdenPeer;
+        auto reply{shardStore->getShardInfo()->makeMessage(app_)};
+        if (peerChainSz > 0)
+            *(reply.mutable_peerchain()) = m->peerchain();
+        send(std::make_shared<Message>(reply, protocol::mtPEER_SHARD_INFO_V2));
+    }
 
-        m->set_hops(m->hops() - 1);
-        if (m->hops() == 0)
-            m->set_lastlink(true);
+    if (m->relays() == 0)
+        return;
 
```

Comment:
It seems whenever a node's shard information is changed, `DatabaseShardImp::updatePeers()` is call to update peers. If so, then `if (m->relays() == 1)`, we can reply with the `ShardInfo`s in `PeerImp::shardInfos_`, and don't have to relay to the last hop which is the dominating cost.